Uses idn_to_* functions to ensure domain names are normalized correctly

Sesquipedalian opened the pull request:
Fixes #7025

 The main thing to fix the bug is just to [use `idn_to_ascii()` and `idn_to_utf8()`](https://github.com/SimpleMachines/SMF2.1/commit/a605ee951b2e0ed03933df70f766d8829c655656) to ensure that domain names containing non-ASCII characters are normalized to their canonical forms (e.g. `t` `a` `¨` `s` `t` `.` `d` `e`  is normalized to  `t` `ä` `s` `t` `.` `d` `e`).

Why do we need to do this? The IANA imposes such normalization of non-ASCII characters for all registered domain names in order to reduce the possibilities for ambiguous domain names, which means that `t` `a` `¨` `s` `t` `.` `d` `e` can never be registered as a domain name—only  `t` `ä` `s` `t` `.` `d` `e` can. This means that if one of our users types `h` `t` `t` `p` `:` `/` `/` `t` `a` `¨` `s` `t` `.` `d` `e` into a post, we need to perform the same normalization on the domain name in our generated links, or else we will end up pointing users to URLs that don't exist.

... But I ran into a further problem: `idn_to_ascii()` and `idn_to_utf8()` are part of the [intl](https://www.php.net/manual/en/book.intl.php) extension, and many PHP installs do not have that extension enabled. So I created [polyfills for them](https://github.com/SimpleMachines/SMF2.1/commit/3883d0264aabafa2a23df3a3f809a4903ee812ab) in Subs-Compat.php. I didn't bother making these do anything more than what we actually need, so they're pretty short.

... But I ran into a further problem: the native `idn_to_ascii()` function internally calls the Unicode Normalization algorithms that are otherwise exposed via the `normalizer_normalize()` function of the intl extension's [Normalizer](https://www.php.net/manual/en/class.normalizer.php) class. But of course, if the intl extension is unavailable then `normalizer_normalize()` is also unavailable, so I had to create a [polyfill for that](https://github.com/SimpleMachines/SMF2.1/commit/eeea8e0e1752178d8264ef79c0f808a0d28dacf8), too. This one needed to be a complete polyfill, so it is. In this case, it seemed best to create functions in Subs-Charset.php that would serve as wrappers for `normalizer_normalize()` if it is available, or that would implement the Unicode Normalization algorithms themselves if `normalizer_normalize()` is not available.

... But I ran into one more problem: my `normalizer_normalize()` polyfills use `mb_ord()` and `mb_chr()` at a couple of key points, but `mb_ord()` and `mb_chr()` were only added in PHP 7.2.0, so I needed to create [polyfills for them](https://github.com/SimpleMachines/SMF2.1/commit/0b7763a1fc8923f0c28b136b208d00ac47203f57) in Subs-Compat.php, too. These are also complete polyfills.

Now, before anyone gets worried about large PRs, let me point out two things:
1. Almost all of this PR is just polyfills, not changes to SMF's behaviour. **Only a605ee9 actually changes any behaviour.** If you have a PHP system with the intl extension enabled, you can test that lone commit without involving any of the other commits at all.
2. I have delicious, nutritious **unit tests** for the polyfills right here: 
    - [Unit testing for SMF's mb_ord() and mb_char() polyfills](https://gist.github.com/Sesquipedalian/0f8ab97b2b73d68d16eaa8fae56f8ed6)
    - [Unit testing for SMF's normalizer_normalize() polyfill](https://gist.github.com/Sesquipedalian/b0870a303967002adc9db93fc48cd919)
    - For the idn_to_* polyfills, just put the list of URLs [here](https://github.com/SimpleMachines/SMF2.1/pull/6754#issuecomment-846471296) into a post directly in SMF.

-----

Also fixes #7046

Sesquipedalian committed:
Implements polyfills for mb_ord() and mb_chr()

Prerequisite for polyfills for idn_to_ascii() and idn_to_utf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Implements wrappers/polyfill for intl::normalizer_normalize()

Prerequisite for polyfills for idn_to_ascii() and idn_to_utf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Implements polyfills for intl::idn_to_ascii() and intl::idn_to_utf8()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Uses idn_to_* functions to ensure domain names are normalized correctly

Fixes #7025

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

live627 commented:
Thanks for all the tests. I just started testing, and in doing so, I noticed that the  third reproduction step did not get transformed at all.

Sesquipedalian commented:
> Thanks for all the tests. I just started testing, and in doing so, I noticed that the third reproduction step did not get transformed at all.

To clarify, when you say "the third reproduction step" do you mean this part?

>    * For the idn_to_* polyfills, just put the list of URLs [here](https://github.com/SimpleMachines/SMF2.1/pull/6754#issuecomment-846471296) into a post directly in SMF.

If so, I'm not sure what you mean. Here's the result in my tests:
![Screen Shot 2021-09-03 at 12 07 41 PM](https://user-images.githubusercontent.com/191731/132048924-7776249f-a817-4ff3-961f-f18ddfe8e99e.png)
You need to inspect the source to see the Punycode host name in the href attribute for the relevant links.

Sesquipedalian committed:
Fixes stupid copy-paste error

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Uses escaped bytes for UTF-8 chars in Subs-Charset.php

This is just to prevent RTL text from causing display issues on GitHub and in some text editors.
Signed-off-by: Jon Stovell <jonstovell@gmail.com>

live627 commented on Sources/Subs-Charset.php:
```diff
@@ -550,4 +550,7794 @@ function utf8_strtoupper($string)
+		if ($string[$c] >= "\xE1\x84\x80" && $string[$c] <= "\xE1\x84\x92" && $string[$c + 1] >= "\xE1\x85\xA1" && $string[$c + 1] <= "\xE1\x85\xB5")
+		{
+			if (!function_exists('mb_ord'))
+				require_once($sourcedir . '/Subs-Compat.php');
```
Undefined global

live627 commented on Sources/Subs-Charset.php:
```diff
@@ -550,4 +550,7794 @@ function utf8_strtoupper($string)
+/**
+ * Helper function for utf8_normalize_c and utf8_normalize_kc
+ */
+function utf8_compose($string)
+{
```
```suggestion
function utf8_compose($string)
{
	global $sourcedir;
```

Sesquipedalian replied:
Good catch

Sesquipedalian committed:
Adds missing global

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Corrects the output of utf8_strto* to match mb_strto*. Fixes #7046

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Uses correct mask value in mb_encode_numericentity $convmap param
This didn't actually cause any bugs, but we may as well be correct.

Signed-off-by: Jon Stovell <jonstovell@gmail.com>
